src/raftstore/store/util.rs

```diff
@@ -107,10 +107,42 @@ pub fn check_key_in_region(key: &[u8], region: &metapb::Region) -> Result<()> {
     }
 }
 
+/// `is_first_vote_msg` checks `msg` is the first vote (or prevote) message or not. It's used for
+/// when the message is receiving but there is no such region in `Store::region_peers` and the
+/// region overlaps with others. In this case we should put `msg` into `pending_votes` instead of
+/// create the peer.
 #[inline]
-pub fn is_first_vote_msg(msg: &RaftMessage) -> bool {
-    msg.get_message().get_msg_type() == MessageType::MsgRequestVote
-        && msg.get_message().get_term() == peer_storage::RAFT_INIT_LOG_TERM + 1
+pub fn is_first_vote_msg(msg: &eraftpb::Message) -> bool {
+    match msg.get_msg_type() {
+        MessageType::MsgRequestVote | MessageType::MsgRequestPreVote => {
+            msg.get_term() == peer_storage::RAFT_INIT_LOG_TERM + 1
+        }
+        _ => false,
+    }
+}
+
+#[inline]
+pub fn is_vote_msg(msg: &eraftpb::Message) -> bool {
+    let msg_type = msg.get_msg_type();
+    msg_type == MessageType::MsgRequestVote || msg_type == MessageType::MsgRequestPreVote
+}
+
+/// `msg_can_create_peer` tells which messages can create new peers.
+// There could be two cases:
+// 1. Target peer already exists but has not established communication with leader yet
+// 2. Target peer is added newly due to member change or region split, but it's not
+//    created yet
+// For both cases the region start key and end key are attached in RequestVote and
+// Heartbeat message for the store of that peer to check whether to create a new peer
+// when receiving these messages, or just to wait for a pending region split to perform
+// later.
+#[inline]
+pub fn msg_can_create_peer(msg: &eraftpb::Message) -> bool {
```

Comment:
Any tests to cover it?

Reply:
How about `is_initial_message`? These messages are the messages that a leader/candidate will send to a unknown peer. Message can't create a peer by themselves, only store can do it.